pkg/sql/alter_primary_key.go

```diff
@@ -327,3 +325,20 @@ func (p *planner) AlterPrimaryKey(
 
 	return nil
 }
+
+// We only recreate the old primary key of the table as a unique secondary
+// index if:
+// * The table has a primary key (no DROP PRIMARY KEY statements have
+//   been executed).
+// * The primary key is not the default rowid primary key.
+// * The new primary key isn't the same hash sharded primary key with a
+//   different bucket count.
+func shouldCopyPrimaryKey(desc *MutableTableDescriptor, newPK *sqlbase.IndexDescriptor) bool {
+	oldPK := desc.PrimaryIndex
+	return desc.HasPrimaryKey() && !desc.IsPrimaryIndexDefaultRowID() &&
+		!(oldPK.IsSharded() && newPK.IsSharded() &&
+			// The first column in the columnIDs is the shard column, which will be different.
```

Comment:
this was a teensy bit hard to read but makes sense after squinting. 

wondering if breaking down to the `if` statements might be better... but it's personal taste for me.

````

func shouldCopyPrimaryKey(desc *MutableTableDescriptor, newPK *sqlbase.IndexDescriptor) bool {
  // no primary key means we have nothing to copy
  if !desc.HasPrimaryKey() {
     return false
  }
  // we never want to copy default row id
  if desc.IsPrimaryIndexDefaultRowID() {
    return false
  }
  if oldPK.IsSharded() && newPK.IsSharded() {
     // ...
  }
  return true
}
```

Reply:
I'm down for this, I find it hard to read as well.